Fix Cooldown for RaiseDead (Ghoul)

tkrokli opened the pull request:
Fix written by @joschiwald on 31 Jan 2014
Fixes issue #11538

Commit:
Fix Cooldown for RaiseDead (Ghoul)

Fix written by joschiwald on 31 Jan 2014
Fixes issue #11538

joschiwald commented:
iirc the code was wrong

tkrokli commented:
OK, I'm still in testing mode, so the PR may still be subject to change. Suggestions are welcome.

Goatform commented:
SMSG_SPELL_GO issue

joschiwald commented:
No, the ghul is summoned by 2 spells
if he dies, there are sent cooldown events of both spell on retail, but tc sends only one, so the other spell remains on inf cooldown on clientside

tkrokli commented:
Yes, I thought your code would fix that problem. At least some users say it did work for them.

Nyeriah commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -22138,6 +22138,34 @@ void Player::SendCooldownEvent(SpellInfo const* spellInfo, uint32 itemId /*= 0*/
+                if (i->first == spellInfo->Id) // skip main spell, already handled above
+                    continue;
+
+                SpellInfo const* spellInfo2 = sSpellMgr->GetSpellInfo(i->first);
```
Just leaving it as a note as @w1sht0l1v3  commented on the issue way long ago, this returns null pointers sporadically so it needs to be checked against that. Otherwise you may experience random crashes.

tkrokli commented:
Thanks for the note, I understand that this can be a problem. I just don't know enough to write a solution to that specific problem. Any pointers to related code are welcome regarding this as well. :smirk_cat:

ccrs commented:
@Nyeriah 

````c++
uint32 cat = spellInfo->GetCategory();
    uint32 catRec = spellInfo->CategoryRecoveryTime;
    if (cat && catRec)
    {
        SpellCategoryStore::const_iterator ct = sSpellsByCategoryStore.find(cat);
        if (ct != sSpellsByCategoryStore.end())
        {
            SpellCategorySet const& catSet = ct->second;
            for (SpellCooldowns::const_iterator i = m_spellCooldowns.begin(); i != m_spellCooldowns.end(); ++i)
            {
                if (i->first == spellInfo->Id) // skip main spell, already handled above
                    continue;

                if (SpellInfo const* spellInfo2 = sSpellMgr->GetSpellInfo(i->first))
                {
                    if (!spellInfo2->IsCooldownStartedOnEvent())
                        continue;

                    if (catSet.find(i->first) != catSet.end())
                    {
                        // Send activate cooldown timer (possible 0) at client side
                        WorldPacket data(SMSG_COOLDOWN_EVENT, 4 + 8);
                        data << uint32(i->first);
                        data << uint64(GetGUID());
                        SendDirectMessage(&data);
                    }
                }
            }
        }
    }
````

Nyeriah commented:
It should look the way you've posted (the spellinfo inside an IF)

tkrokli commented:
> Spells in that context may not always be valid, and when they aren't, IsCooldownStartedOnEvent() will crash the server.

I see. That would be unstable and unwanted in the source. Does it mean that the whole code section (originally written by joschiwald) needs to be replaced with a different solution?

ariel- commented:
Change

```c++
if (!spellInfo2->IsCooldownStartedOnEvent())
    continue;
```

with

```c++
if (!spellInfo2 || !spellInfo2->IsCooldownStartedOnEvent())
    continue;
```

Commit:
Update Player.cpp

Added code suggested by @ariel-

tkrokli commented:
OK, I have added that change. Now I need to "squash" my commits, but I don't know how to apply that command directly to my fork of the github.com tree. Previously, I had to remove my current clone and fork a new clone of the github.com/TrinityCore/TrinityCore/tree/3.3.5/ branch and then reapply the changes to the file I had changed multiple times before. (Sorry for my lack of experience with merges.)

DDuarte commented:
This might help: https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request#squash-your-changes-optional

tkrokli commented:
> This might help: https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request#squash-your-changes-optional

Thanks for the link. I tried following it, but either it made no changes to my fork or it tried to push to the TrinityCore repo (but got a 403 error, of course). I need to go with what works for me.

tkrokli commented:
Sorry about the mess, I have to close this PR and create a new one based on the updated source.

tkrokli commented:
:+1: Thanks to all who have added to this PR, even though I closed it because I deleted my repo.
I am currently learning how to use Git Bash (instead of GitExt or Git GUI) so I can base, rebase, pick, squash and push changes to my own repository. If you are still interested in this issue, head over to https://github.com/TrinityCore/TrinityCore/pull/13941 to see if I have made any useful progress or if I need your feedback to correct my current PR. :+1:
